edb/pgsql/compiler/dml.py

```diff
@@ -245,40 +245,68 @@ def gen_dml_cte(
     ctx: context.CompilerContextLevel,
 ) -> tuple[pgast.CommonTableExpr, pgast.PathRangeVar]:
 
-    target_ir_set = ir_stmt.subject
-    target_path_id = target_ir_set.path_id
-
-    relation = relctx.range_for_typeref(
-        typeref,
-        target_path_id,
-        for_mutation=True,
-        ctx=ctx,
-    )
-    assert isinstance(relation, pgast.RelRangeVar), (
-        "spurious overlay on DML target"
-    )
+    subject_ir_set = ir_stmt.subject
+    subject_path_id = subject_ir_set.path_id
 
     dml_stmt: pgast.InsertStmt | pgast.SelectStmt | pgast.DeleteStmt
+    subject_rvar: pgast.BaseRangeVar
     if isinstance(ir_stmt, irast.InsertStmt):
+        relation = relctx.range_for_typeref(
+            typeref,
+            subject_path_id,
+            for_mutation=True,
+            ctx=ctx,
+        )
+        assert isinstance(relation, pgast.RelRangeVar), (
+            "spurious overlay on DML target"
+        )
+
         dml_stmt = pgast.InsertStmt(relation=relation)
+        subject_rvar = relation
+
     elif isinstance(ir_stmt, irast.UpdateStmt):
         # We generate a Select as the initial statement for an update,
         # since the contents select is the query that needs to join
         # the range and include policy filters and because we
         # sometimes end up not needing an UPDATE anyway (if it only
         # touches link tables).
         dml_stmt = pgast.SelectStmt()
+
+        subject_rel_overlayed = relctx.range_for_typeref(
+            typeref,
+            subject_path_id,
+            for_mutation=False,
+            include_descendants=False,
+            dml_source=[
+                k for k in ctx.dml_stmts.keys()
+                if isinstance(k, irast.MutatingLikeStmt)
+            ],
```

Comment:
Do we want to only do this in SQL mode, though?

We should also document why we do this.

Reply:
We could yeah, but I don't have a nice variable close by that would indicate if this is "SQL mode" or not.